### PR TITLE
First steps with contracts

### PR DESCRIPTION
**Pull request by sububer**
**UPDATE**

Here's some auction/marketplace contracts, an interactive js script, and an ERC721 and ERC1155 contract examples.

Main points:

`Auction.sol` -- my  first stab and definiing auction logic, but not sure if it is right for how wer're thinking of listing/viewing aucitons

`G6Marketplace.sol` is more of an experiment making and buying nft items -- no auction.. just listing/buying.

`SimpleAuction.sol` is just that... creating an auction by instantiating the auction, then can bid, auctionEnd, and withdrawl

also found this guy on youtube -- which is smilar in ways to the above, but maybe we can grab from it.
`MarketplaceBoilerplate.sol`

https://youtu.be/WZWCzsB1xUE
https://raw.githubusercontent.com/ethereum-boilerplate/ethereum-nft-marketplace-boilerplate/main/src/contracts/marketplaceBoilerplate.sol

Two contracts:
G6NFT - erc721
GameItems. - erc1155

Added a couple of contracts to learn how this repo works for compiling and deploying, as well as to start thinking about ERC1155 and the general mechanics of an auction contract.

`Auction.sol` is a contract which handles the owner, duration, bidding, winning and ability to withdrawl after an auction. I tested the logic locally and it seems ok. Its not in the right ERC1155 spec.. I was just seeing how I could define an auction.

`GameItems.sol` is taken from a hello world ERC1155 contract from here:
https://docs.openzeppelin.com/contracts/4.x/erc1155

uses the helper interface/module from OpenZeppelin to get all the functionality. I haven't done much with it other than compile and deploy it.

I was able to compile and deploy both of those. I'd like to discuss more about the right way to setup the interface correctly for the auction.. maybe later this week.

Some good resources I've found:

https://docs.openzeppelin.com/contracts/4.x/api/token/erc1155
https://ethereum.org/en/developers/docs/standards/tokens/erc-1155/

another reference implementation:
https://github.com/enjin/erc-1155/blob/master/contracts/ERC1155.sol

Doesn't matter if we merge this or not, but I wanted to share.

### CODE REVIEW

File: contracts/Auction.sol [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+        bids[msg.sender] = currentBid;
+
+        // is this bid now highest biding, and greater than current highest bidder?
+        if(currentBid <= bids[highestBidder]){
+            highestBindingBid = min(currentBid + bidIncrement, bids[highestBidder]);
+        } else {
+            highestBindingBid = min(currentBid, bids[highestBidder] + bidIncrement);
+            highestBidder = payable(msg.sender);
+        }
```
Comment (pmm09c) [inline contracts/Auction.sol:92]: generally, you want to issue an event so off-chain indexing tools can update on events, if no event is declared you have no way of knowing unless you're polling a public variable.

Comment (pmm09c) [reply contracts/Auction.sol:92]: same comments for all functions

File: contracts/Auction.sol [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    string public ipfsHash;
+
+    enum State {Started, Running, Ended, Canceled}
+    State public auctionState;
+
+    uint public highestBindingBid;
+    address payable public highestBidder;
+
+    // keeping addresses, amounts of bidders in a map
+    mapping (address => uint) public bids;
```
Comment (pmm09c) [inline contracts/Auction.sol:22]: what happens if one user bids on multiple items

**Review by pmm09c — APPROVED** [review]

File: contracts/Auction.sol (hunk 2 above)
Comment (sububer) [reply contracts/Auction.sol:22]: right.. my mental model for this exercise was that an auction was one entity with an owner for one item/batch per auction. obv.. need to discuss how this really should be modeled.

File: contracts/Auction.sol (hunk 1 above)
Comment (sububer) [reply contracts/Auction.sol:92]: I follow what you're saying as it makes sense. Do you have any concrete examples to refer to? I'm used to data-flow, or event-streams with traditional api backend architecture, but i'm not sure how to map that to this evm/contract/dapp space yet.

File: contracts/Auction.sol (hunk 2 above)
Comment (pmm09c) [reply contracts/Auction.sol:22]: https://www.tutorialspoint.com/solidity/solidity_events.htm

File: contracts/Auction.sol (hunk 1 above)
Comment (pmm09c) [reply contracts/Auction.sol:92]: https://www.tutorialspoint.com/solidity/solidity_events.htm

Comment (pmm09c) [reply contracts/Auction.sol:92]: https://thegraph.com/docs/en/about/introduction/ This explains one way to pull the event data off-chain for indexing so the web2 side can utilize the information

Comment (sububer) [reply contracts/Auction.sol:92]: Cool... taking a look.. thanks!

**Pull request merged**